src/Analyser/MutatingScope.php

```diff
@@ -506,6 +506,13 @@ public function hasVariableType(string $variableName): TrinaryLogic
 	/** @api */
 	public function getVariableType(string $variableName): Type
 	{
+		if (!$this->treatPhpDocTypesAsCertain) {
+			if ($variableName === 'this' && $this->hasVariableType('this')->yes()) {
```

Comment:
Not exactly correct. 'this' should be duplicated for nativeTypes where assigned to variableType 'this'